okhttp/src/main/java/io/grpc/okhttp/OkHttpServerTransport.java

```diff
@@ -164,15 +165,17 @@ private void startIo(SerializingExecutor serializingExecutor) {
       // The socket implementation is lazily initialized, but had broken thread-safety 
       // for that laziness https://bugs.openjdk.org/browse/JDK-8278326. 
       // As a workaround, we lock to synchronize initialization with shutdown().
+      HandshakerSocketFactory.HandshakeResult result;
       synchronized (lock) {
         socket.setTcpNoDelay(true);
       }
-      HandshakerSocketFactory.HandshakeResult result =
-          config.handshakerSocketFactory.handshake(socket, Attributes.EMPTY);
+      synchronized (handShakeLock) {
+        result = config.handshakerSocketFactory.handshake(socket, Attributes.EMPTY);
+      }
       synchronized (lock) {
         this.socket = result.socket;
+        this.attributes = result.attributes;
```

Comment:
Did TSAN show this was needed? This field shouldn't be looked at until `listener.transportReady()`. I'd be interested to see what TSAN found.